sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataOutboundAggregator.java

```diff
@@ -0,0 +1,333 @@
+  private final Map<KV<String, String>, ByteString.Output> timersBuffer;
+  @VisibleForTesting ScheduledFuture<?> flushFuture;
+  private Long totalByteCounter = 0L;
+  private final Object flushLock = new Object();
+
+  public BeamFnDataOutboundAggregator(
+      PipelineOptions options,
+      Supplier<String> processBundleRequestIdSupplier,
+      StreamObserver<BeamFnApi.Elements> outboundObserver) {
+    this.sizeLimit = getSizeLimit(options);
+    this.timeLimit = getTimeLimit(options);
+    this.outputDataCoders = new HashMap<>();
+    this.outputTimersCoders = new HashMap<>();
+    this.perEndpointDataByteCount = new HashMap<>();
+    this.perEndpointTimersByteCount = new HashMap<>();
+    this.perEndpointElementCount = new HashMap<>();
+    this.outboundObserver = outboundObserver;
+    this.dataBuffer = new HashMap<>();
+    this.timersBuffer = new HashMap<>();
+    if (timeLimit > 0) {
+      this.flushFuture =
+          Executors.newSingleThreadScheduledExecutor(
+                  new ThreadFactoryBuilder()
+                      .setDaemon(true)
+                      .setNameFormat("DataBufferOutboundFlusher-thread")
+                      .build())
+              .scheduleAtFixedRate(
+                  this::periodicFlush, timeLimit, timeLimit, TimeUnit.MILLISECONDS);
+    } else {
+      this.flushFuture = null;
+    }
+    this.processBundleRequestIdSupplier = processBundleRequestIdSupplier;
+  }
+
+  /**
+   * Register the outbound data logical endpoint for the transform alongside it's output value
+   * coder.
+   */
+  public <T> FnDataReceiver<T> registerOutputDataLocation(String pTransformId, Coder<T> coder) {
+    outputDataCoders.put(pTransformId, coder);
+    if (timeLimit > 0) {
+      return data -> {
+        synchronized (flushLock) {
+          accept(
+              dataBuffer.computeIfAbsent(pTransformId, id -> ByteString.newOutput()),
+              (Coder<T>) outputDataCoders.get(pTransformId),
+              perEndpointDataByteCount.computeIfAbsent(pTransformId, id -> new LongAdder()),
+              data);
+        }
+      };
+    }
+    return data ->
+        accept(
+            dataBuffer.computeIfAbsent(pTransformId, id -> ByteString.newOutput()),
+            (Coder<T>) outputDataCoders.get(pTransformId),
+            perEndpointDataByteCount.computeIfAbsent(pTransformId, id -> new LongAdder()),
+            data);
+  }
+
+  /**
+   * Register the outbound timers logical endpoint for the transform alongside it's output value
+   * coder.
+   */
+  public <T> FnDataReceiver<T> registerOutputTimersLocation(
+      String pTransformId, String timerFamilyId, Coder<T> coder) {
+    KV<String, String> timerKey = KV.of(pTransformId, timerFamilyId);
+    outputTimersCoders.put(timerKey, coder);
+    if (timeLimit > 0) {
+      return timers -> {
+        synchronized (flushLock) {
+          accept(
+              timersBuffer.computeIfAbsent(timerKey, key -> ByteString.newOutput()),
```

Comment:
You'll want to turn this into a private class so you can store these values as member variables instead of doing multiple map lookups on each accept invocation or during flushing. Also consider storing maps of these _metric tracking_ receivers instead of maps for `coders`, `adders`, `bytestring.output`, ...
Then this code can look like:
```
synchronized (flushLock) {
  FnDataReceiver<T> receiver = new Receiver<T>(coder);
  timerReceivers.put(timerKey, receiver);  // should be an error to overwrite an existing registration
  return timers -> { synchronized(flushLock) { receiver.accept(timers); } };
}
```
The unsynchronized code below would just return the `Receiver` instance without wrapping the construction or execution with synchronized.

Reply:
Done.